Source/Core/InputCommon/ControllerInterface/evdev/evdev.cpp

```diff
@@ -0,0 +1,270 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <libudev.h>
+#include <unistd.h>
+
+#include "Common/Logging/Log.h"
+#include "InputCommon/ControllerInterface/evdev/evdev.h"
+
+
+namespace ciface
+{
+namespace evdev
+{
+
+void Init(std::vector<Core::Device *> &controllerDevices)
+{
+	int num_controllers = 0;
+
+	// We use Udev to find any devices. In the future this will allow for hotplugging.
+	// But for now it is essentially iterating over /dev/input/event0 to event31. However if the
+	// naming scheme is ever updated in the future, this *should* be fowards compatable.
+
+	struct udev *udev = udev_new();
+	_assert_msg_(PAD, udev != 0, "Couldn't initilize libudev.");
+
+	// List all input devices
+	struct udev_enumerate *enumerate = udev_enumerate_new(udev);
+	udev_enumerate_add_match_subsystem(enumerate, "input");
+	udev_enumerate_scan_devices(enumerate);
+	struct udev_list_entry *devices = udev_enumerate_get_list_entry(enumerate);
+
+	// Iterate over all input devices
+	struct udev_list_entry *dev_list_entry;
+	udev_list_entry_foreach(dev_list_entry, devices)
+	{
+		const char *path = udev_list_entry_get_name(dev_list_entry);
+
+		struct udev_device *dev = udev_device_new_from_syspath(udev, path);
+
+		const char *devnode = udev_device_get_devnode(dev);
+		// We only care about devices which we have read/write access to.
+		if(access(devnode, W_OK) == 0)
+		{
+			// Unfortunately udev gives us no way to filter out the non event device interfaces.
+			// So we open it and see if it works with evdev ioctls or not.
+			evdevDevice *input = new evdevDevice(devnode, num_controllers);
+
+			if(input->isInteresting())
+			{
+				controllerDevices.push_back(input);
+				num_controllers++;
+			}
+			else
+			{
+				// Either it wasn't a evdev device, or it didn't have at least 8 buttons or two axis.
+				delete input;
+			}
+		}
+		udev_device_unref(dev);
+	}
+	udev_enumerate_unref(enumerate);
+	udev_unref(udev);
+}
+
+evdevDevice::evdevDevice(std::string devnode, int id) : m_devfile(devnode), m_id(id)
+{
+	// The device file will be read on one of the main threads, so we open in non-blocking mode.
+	m_fd = open(devnode.c_str(), O_RDWR|O_NONBLOCK);
+	int ret = libevdev_new_from_fd(m_fd, &m_dev);
+
+	if(ret != 0)
+	{
+		// This useally fails because the device node isn't an evdev device, such as /dev/input/js0
+		m_initilized = false;
+		close(m_fd);
+		return;
+	}
+
+	const char *name = libevdev_get_name(m_dev);
+	m_name = name ? std::string(name) : "Unknown Device";
```

Comment:
Yeah, it's all the other functions that can potentially (and do in practice) return null.